openedx_learning/core/components/models.py

```diff
@@ -319,7 +319,43 @@ class Content(models.Model):
     # be UTC.
     created = manual_date_time_field()
 
-    data = models.BinaryField(null=False, max_length=MAX_SIZE)
+    # There are two ways to store the raw content. You can choose to use one or
+    # both, with the following tradeoffs:
+    #
+    # data: Use this for smaller content that needs to be accessed quickly and
+    #       and examined by the server-side code. Content from this field is not
+    #       meant to be directly exposed to browsers.
+    #
+    # file: Use this for data that needs to be downloaded by a browser, or for
+    #       files that are just much larger.
+    #
+    # If you decide to use both fields (e.g. for a small file that you want to
+    # read in Python code but also make available to browsers), the values must
+    # match.
+    #
+    # If the content being stored is text, it is strongly recommended to use
+    # utf-8 encoding (without BOM).
+    #
+    # Ramblings:
+    # 
+    # Have 2 fields for now:
+    #   - text
+    #   - file
+    #   - 
+    # If it becomes an actual use case to have locally processed, low-latency
+    # binary files (e.g. wasm) later, then add a new nullable field for binary.
+    # 
+    # Add a separate toggle for whether it's publicly accessible? So the only
+    # things that can be accessed by browser are things that are marked publicly
+    # accessible and have file attribute. So they can't download python_lib.zip
+    # and possibly other sensitive things.
+    #
+    # Actually, access is probably controlled at the ComponentVerison level?
+    # That could get problematic.
```

Comment:
Hm, interesting. How about cases where it should only be accessible to some subset of logged-in users? (eg, a graphic that only paid+enrolled users in a course should be able to download).

I assume that logic can't live in MediaServer, since MediaServer needs to be interchangeable with a CDN.

Signed URLs, maybe?

Reply:
Yeah, definitely signed URLs for when it's private-download. I'm hoping we don't have to figure out full fledged roles/permissions just yet, but I think there are two, maybe three types of assets:

1. Publicly downloadable/cacheable in CDN.
2. Private by default, but signed URLs can be generated for downloading.
3. (Possibly) Never downloadable in the LMS (e.g. grader code).

Reply:
Cool. 1 and 2 make sense.

3 is interesting, and it's hard for me to know whether it'd be useful or redundant. I could see it being a nice safeguard against us generating a signed URL for something we shouldn't. Without seeing the signing code, though, it's hard to know. You could always add it now, and drop the column later if it doesn't prove to be useful.

Reply:
Dave and I chatted about this. To summarize: We're avoiding distinguishing between (1) and (2) at this layer because that distinction is controlled by others layers (eg, course visibility settings), and changes to it wouldn't necessarily be captured by a version change. Thus, we couldn't count on the value of `is_public` to maintain correctness. The distinction between (1/2) and (3) (that is: `is_private` aka `not learner_downloadable`), on the other hand, is something that can be determined entirely from the asset's usage in a particular ComponentVersion.

Reply:
Expanded the comment to try to capture all this.